qiskit/transpiler/passes/remove_diagonal_gates_before_measurere.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+
+"""
+Transpiler pass to remove diagonal gates (like RZ, T, Z, etc) before
+a measurement. Including diagonal control gates.
+"""
+
+from qiskit.circuit import Measure
+from qiskit.extensions.standard import RZGate, ZGate, TGate, SGate, TdgGate, SdgGate, U1Gate,\
+    CzGate, CrzGate, Cu1Gate, RZZGate
+from qiskit.transpiler.basepasses import TransformationPass
+
+
+class RemoveDiagonalGatesBeforeMeasure(TransformationPass):
+    """Remove diagonal gates (like RZ, T, Z, etc) before a measurement.
+    Including diagonal control gates."""
+
+    def run(self, dag):
+        """Return a new circuit that has been optimized."""
+        diagonal_gates = (RZGate, ZGate, TGate, SGate, TdgGate, SdgGate, U1Gate)
+        diagonal_control_gates = (CzGate, CrzGate, Cu1Gate, RZZGate)
+
+        nodes_to_remove = set()
+        for measure in dag.op_nodes(Measure):
+            for predecessor in dag.predecessors(measure):
```

Comment:
There is a problem here. You should not traverse over all predecessors of a `measure`. Each `measure` node has exactly two predecessors, one on the qubit edge and one on the clbit edge. Here we should *only* be concerned about the qubit predecessor. Otherwise nodes may wrongly be removed. I'll explain more in a comment.

Reply:
Awesome catch.  It is easily solvable with a `quantum_predecessors`, which does not exists in the current DAG API. However, there is a `quantum_successors`, so I think is natural to introduce it. That is done in 019ca0be (with the test).

I'm using that new feature in the DAG API to fix the issue in 11a0a743.